Bottom: becb42a7b520e4bc05fffb8f9dece5e304ccc2f8
Top:    bacecbccffd0e6c0a72011fccb8a4f17c2261f83
Author: Jason Wang <jasowang@redhat.com>
Date:   2017-03-02 12:10:10 +0800

vhost: partially open code get_rx_bufs()

To simplify the batching experiments.

Signed-off-by: Jason Wang <jasowang@redhat.com>


---

diff --git a/drivers/vhost/net.c b/drivers/vhost/net.c
index 15d689b..c609d5a 100644
--- a/drivers/vhost/net.c
+++ b/drivers/vhost/net.c
@@ -710,11 +710,19 @@ static void handle_rx(struct vhost_net *net)
 			headcount = get_rx_bufs(vq, vq->heads, vhost_len,
 						&in, vq_log, &log, UIO_MAXIOV);
 		} else {
-			headcount = get_rx_bufs(vq, vq->heads, 1,
-						&in, vq_log, &log, 1);
-			if (headcount > 0) {
-				vhost_len = vq->heads[0].len;
+			unsigned int out;
+			headcount = vhost_get_vq_desc(vq, vq->iov,
+						ARRAY_SIZE(vq->iov),
+						&out, &in, vq_log, &log);
+			if (headcount == vq->num)
+				headcount = 0;
+			else if (headcount >= 0) {
+				vq->heads[0].id = cpu_to_vhost32(vq, headcount);
+				vhost_len = vq->heads[0].len =
+					cpu_to_vhost32(vq,
+						iov_length(vq->iov, in));
 				sock_len = vhost_len - vhost_hlen;
+				headcount = 1;
 			}
 		}
